### ISSUE TITLE
Experimental distributed SGD API for Ray (e.g., ray.train)

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#1945]
### Describe the problem

We should provide a distributed SGD API as a part of Ray. This should be usable both for large-scale training (e.g., ImageNet), and also for RLlib (improving upon the existing multi-gpu optimizer). It should scale to 64-128 GPUs and provide performance competitive with Distributed TF and Horovod, but integrate well within the Ray APIs.

There are a couple strategies for distributed SGD that we can provide
- ring allreduce (and other allreduce algorithms)
- sharded parameter server

In addition, we will need to include a custom TensorFlow op to allow TensorFlow to read/write plasma memory directly within a TF run. This will also allow for the pipelined transmission of gradients through plasma.

Current status: we are prototyping this, and hope to have performance numbers and a more detailed design doc ready in a few weeks. This project depends on the x-ray internal refactoring to improve Ray performance.

**Comment by ericl** [comment]
FYI @joneswong

**Comment by richardliaw** [comment]
One common issue I'm running into: on P3 instances, I'm sometimes unable to access the GPU. I can run `nvidia-smi` fine, but somehow TF can't access the GPUs.

```
(tensorflow_p27) ubuntu@ip-172-31-80-212:~/osdi2018$ CUDA_VISIBLE_DEVICES=0,1,2,3 ipython
Python 2.7.14 |Anaconda, Inc.| (default, Dec  7 2017, 17:05:42)
Type "copyright", "credits" or "license" for more information.

IPython 5.5.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import resnet101
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ubuntu/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
/home/ubuntu/anaconda3/envs/tensorflow_p27/lib/python2.7/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ubuntu/.config/matplotlib/matplotlibrc", line #3
  (fname, cnt))
***unable to import torch***

In [2]: import tensorflow as tf

In [3]: with tf.device("/gpu:2"):
   ...:     y = resnet101.TFBenchModel()
   ...:
2018-05-18 00:36:02.400875: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
WARNING:tensorflow:From tfbench/convnet_builder.py:372: calling reduce_mean (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead

In [4]: y.device
Out[4]: ''
```

It would be nice to investigate this.

**Comment by robertnishihara** [comment]
Is it just TF? Do PyTorch or other frameworks have this issue (on the same instance)?

**Comment by calclavia** [comment]
It would be nice to have Pytorch support for this, or have some kind of way to use Pytorch's DistributedDataParallel system with Ray's Kubernetes system

**Comment by JonathanChiang** [comment]
This feature seems similar to pysyft? I would love to work on this problem.

**Comment by richardliaw** [comment]
I think PySyft is for private data? This is just a SGD implementation (without security concerns).

@pschafhalter is working on this in #4797. cc @calclavia

**Comment by JonathanChiang** [comment]
From my understanding pysyft is designed for differential privacy. I'm curious then to leverage Ray's architecture with differential privacy across different nodes for distributed deep learning. 

I'm assuming the distributed deep learning is still within ray's scope?? But I'm guessing the differential privacy is not.

**Comment by robertnishihara** [comment]
Done in experimental form.

**Comment by JonathanChiang** [comment]
Can you be more specific about the experimental form or point me to what you are describing?

**Comment by robertnishihara** [comment]
@JonathanChiang some minimal documentation at https://ray.readthedocs.io/en/latest/distributed_training.html. The initial PR was https://github.com/ray-project/ray/pull/4797. Any feedback would be great!
